### PR TITLE
fix: Resolve mid-route swing => deadhead issues in minischedule & swing view

### PR DESCRIPTION
**Pull request by KaylaBrady**
ticket: https://app.asana.com/0/1148853526253437/1203278600620513/f

### CODE REVIEW

File: assets/src/components/propertiesPanel/minischedule.tsx [hunk 1]
```diff
@@ -461,16 +471,30 @@ const MidRouteSwingOnSecondHalf = ({
   />
 )
 
-const getSequence = (tripIndex: number, pieceTrips: (Trip | AsDirected)[]) => {
+type Sequence = "first" | "middle" | "last"
+type NonRevenueStatus = "pull-out" | "pull-back" | "deadhead"
```
Comment (lemald) [inline assets/src/components/propertiesPanel/minischedule.tsx:475]: **praise:** I like splitting these out into separate types even if it's a simple mapping, it makes things a bit clearer.

File: assets/tests/components/propertiesPanel/minischedule.test.tsx [hunk 2]
```diff
@@ -178,6 +178,33 @@ const midRouteSwingPiece2: Piece = {
...
+    },
+  ],
+  endTime: 480,
+  endPlace: "terminal1",
+  startMidRoute: {
+    time: 180,
+    trip: midRouteSwingTrip1,
+  },
+  endMidRoute: false,
+}
```
Comment (lemald) [inline assets/tests/components/propertiesPanel/minischedule.test.tsx:206]: **suggestion:** consider making a factory for this - we use Fishery and there are some existing factories located in `assets/tests/factories`.

File: lib/schedule/swing.ex [hunk 3]
```diff
@@ -116,4 +116,28 @@ defmodule Schedule.Swing do
...
+      |> Enum.take(1)
+      |> List.first()
+
+    case {first_full_trip, piece.start_mid_route?} do
+      {nil, %{trip: mid_route_trip}} ->
+        mid_route_trip |> trip_or_trip_id_to_trip(trips_by_id)
+
+      {_, %{trip: mid_route_trip}} ->
+        with %{} = first_trip <- mid_route_trip |> trip_or_trip_id_to_trip(trips_by_id) do
+          %{first_trip | :run_id => first_full_trip.run_id}
```
Comment (lemald) [inline lib/schedule/swing.ex:136]: You can use the slightly more compact map syntax when the key is an atom.

```suggestion
          %{first_trip | run_id: first_full_trip.run_id}
```

File: lib/schedule/swing.ex [hunk 4]
```diff
@@ -116,4 +116,28 @@ defmodule Schedule.Swing do
...
+    first_full_trip =
+      piece.trips
+      |> Stream.map(&trip_or_trip_id_to_trip(&1, trips_by_id))
+      |> Stream.filter(fn trip -> !is_nil(trip) end)
+      |> Enum.take(1)
+      |> List.first()
+
+    case {first_full_trip, piece.start_mid_route?} do
+      {nil, %{trip: mid_route_trip}} ->
+        mid_route_trip |> trip_or_trip_id_to_trip(trips_by_id)
```
Comment (lemald) [inline lib/schedule/swing.ex:132]: **question:** is this one of the data edge cases you were referring to? If I understand correctly, this is the case in which the piece starts with a swing but has no other trips. I'd say to make the same update you do in the other case to make sure the `run_id` is correct juuuust in case this ever comes up, but I doubt it's likely to ever happen.

Comment (KaylaBrady) [reply lib/schedule/swing.ex:132]: Yeah this is that edge case! In this case though, what would the correct `run_id` be? Since there are no other full trips, would it be just the `run_id` of the swing?

Comment (lemald) [reply lib/schedule/swing.ex:132]: I was thinking you could use the `run_id` from the `piece`, provided that that's actually populated and correct.

Comment (KaylaBrady) [reply lib/schedule/swing.ex:132]: Ah! Do you have a sense of how reliable that field is? Wondering if that could be used for the `run_id` so we no longer have to look for `first_full_trip` to get the run_id when there is a mid route swing-on

File: test/schedule/swing_test.exs [hunk 5]
```diff
@@ -333,41 +322,141 @@ defmodule Schedule.SwingTest do
...
+                   to_route_id: ^route_id_during_swing,
+                   to_trip_id: ^trip_id_during_swing,
+                   to_run_id: ^run_id_after_swing
                  }
                ]
-             }
+             } = Swing.from_blocks(blocks, trips_by_id)
+    end
+
+    @tag :skip
```
Comment (lemald) [inline test/schedule/swing_test.exs:363]: **question:** is this intentional? Looks like it may have inadvertently been left in.

File: lib/schedule/swing.ex (hunk 4 above)
Comment (lemald) [reply lib/schedule/swing.ex:132]: I think it should be pretty reliable. If I'm not mistaken that value would get populated [here](https://github.com/mbta/skate/blob/a837fa82f60652464deeae640c8152b04cf6bdff/lib/schedule/hastus/activity.ex#L135), and it appears that it's ultimately using the value of `run_id` from the relevant row in `activities.csv`.

File: test/schedule/swing_test.exs (hunk 5 above)
Comment (firestack) [reply test/schedule/swing_test.exs:363]: that test is failing due to the data edge case, but with your other comments we're able to enable it.

File: assets/tests/components/propertiesPanel/minischedule.test.tsx (hunk 2 above)
Comment (firestack) [reply assets/tests/components/propertiesPanel/minischedule.test.tsx:206]: Experimenting/learning fishery here https://github.com/mbta/skate/tree/kb-kf-minischedule-fix-/fishery-factories

File: assets/src/components/propertiesPanel/minischedule.tsx [hunk 6]
```diff
@@ -323,10 +323,13 @@ const Piece = ({
...
   activeIndex: [number, number] | null
 }) => {
+  const startsWithMidRouteTrip = piece.startMidRoute ? true : false
   const isSwingOn: boolean =
     piece.trips.length > 0 &&
     isTrip(piece.trips[0]) &&
-    !isDeadhead(piece.trips[0])
+    !isDeadhead(
+      startsWithMidRouteTrip ? piece.startMidRoute!.trip : piece.trips[0]
+    )
```
Comment (firestack) [inline assets/src/components/propertiesPanel/minischedule.tsx:332]: this is causing a warning about 
```
  331:32  warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion
```
So I'm considering reverting the change and going back to `piece.startMidRoute ? <a> : <b>`. 

That okay with you @KaylaBrady ?

**Review by lemald — APPROVED** [review]
Nice to see a bunch of new Fishery factories added!

File: assets/src/components/propertiesPanel/minischedule.tsx (hunk 6 above)
Comment (firestack) [reply assets/src/components/propertiesPanel/minischedule.tsx]: addressed in [mbta/skate@`52985e1` (#1843)](https://github.com/mbta/skate/pull/1843/commits/52985e175b13f54c74c74b33fb1bd7f3edd2e87b)
